src/compiler/transformers/es2015.ts

```diff
@@ -1188,6 +1191,190 @@ export function transformES2015(context: TransformationContext): (x: SourceFile
         return block;
     }
 
+    function transformConstructorBodyWorker(
+        prologueOut: Statement[],
+        statementsOut: Statement[],
+        statementsIn: NodeArray<Statement>,
+        statementOffset: number,
+        superPath: readonly number[],
+        superPathDepth: number,
+        constructor: ConstructorDeclaration & { body: FunctionBody },
+        isDerivedClass: boolean,
+        hasSynthesizedSuper: boolean,
+        isFirstStatement: boolean,
+    ): boolean {
+        let mayReplaceThis = false;

```

Comment:
what does mayReplaceThis mean? I can see that transformConstructorBody uses it to decide whether to `return _this` but I don't know why the constructor would or would not need that statement.

Reply:
If your constructor only contains a call to `super()`, we emit it downlevel as this:

```js
return _super.call(this) || this;
```

However, if you have other statements that follow `super()`, we use a `_this` variable that that we use in place of `this` references, in case the superclass returned an object from its constructor. In that case, the return value of the super constructor becomes the new `this`:

```js
var _this = _super.call(this) || this;
// ... lines that may have referenced `this` that now reference `_this`...
return _this;
```

IIRC, a few years ago @DanielRosenwasser introduced some simplifications for the `var _this = ...; return _this;` emit when the `_this` wasn't necessary. This logic is intended to maintain that simplified emit.